File: scraperlibs/scraperwiki/metadata.php
<?php
class SW_MetadataClient
{   
    private static $instance;

    public static function create()
    {
        if (!isset(self::$instance))
        {
            self::$instance = new SW_MetadataClient();
        }

        return self::$instance;
    }

    function __construct()
    {
        if(!getenv("SCRAPER_GUID"))
        {
            $this->metadata_local = Array("title" => "Untitled Scraper", "CPU limit" => "100");
        }

    }

    private function get_url($metadata_name)
    {
        return sprintf("http://%s/scrapers/metadata_api/%s/%s/", getenv("metadata_host"), getenv("SCRAPER_GUID"), urlencode($metadata_name));
    }
    
    private function get_metadata($metadata_name)
    {
        $ch = curl_init();
        curl_setopt($ch, CURLOPT_URL, $this->get_url($metadata_name));
        curl_setopt ($ch, CURLOPT_RETURNTRANSFER, true);
        $res = curl_exec($ch);
        $status = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        curl_close($ch);
        if($status == 200)
        {
            return json_decode($res);
        }
        else
        {
            return null;
        }
    }
    
    public function get($metadata_name, $default=null)
    {
        if(!getenv("SCRAPER_GUID"))
        {
            $value = $this->metadata_local[$metadata_name];
            if($value == null)
            {
                return $default;
            }
            return $value;
        }

        $metadata = $this->get_metadata($metadata_name);
        if($metadata)
        {
            return json_decode($metadata->value);
        }
        else
        {
            return $default;
        }
    }    

    
    public function get_run_id($metdata_name)
    {
        $metadata = $this->get_metadata($metadata_name);
        if($metadata)
        {
            return json_decode($metadata->run_id);
        }
        else
        {
            return null;
        }
    }
    
    public function save($metadata_name, $value)
    {
        if(!getenv("SCRAPER_GUID"))
        {
            print 'The scraper has not been saved yet. Metadata will not be persisted between runs\n';
            $this->metadata_local[$metadata_name] = $value;
        }

        $ch = curl_init();
        curl_setopt($ch, CURLOPT_URL, $this->get_url($metadata_name));
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);

        if($this->get($metadata_name))
        {
            curl_setopt($ch, CURLOPT_CUSTOMREQUEST, 'PUT');
        }
        else
        {
            curl_setopt($ch, CURLOPT_CUSTOMREQUEST, 'POST');
        }

        curl_setopt($ch, CURLOPT_POSTFIELDS, sprintf("run_id=%s&value=%s", getenv("RUNID"), json_encode($value)));
        
        $res = curl_exec($ch);
        curl_close($ch);
        return $res;
    }
}
?>

File: scraperlibs/scraperwiki/metadata.rb
require 'uri'
require 'net/http'
require 'json'
require 'singleton'
require 'cgi'

class LocalMetadataClient
  include Singleton

  def initialize
    @metadata_local = { "title" => "Untitled Scraper", "CPU limit" => "100" }
  end
  
  def get(metadata_name, default = nil)
    value = @metadata_local[metadata_name]
    return value ? value : default
  end
  
  def save(metadata_name,value)
    puts 'The scraper has not been saved yet. Metadata will not be persisted between runs'
    @metadata_local[metadata_name] = value
    return nil
  end
end

class SW_MetadataClient
  include Singleton

  def self.create()
    if ENV["SCRAPER_GUID"]
      SW_MetadataClient.instance()
    else
      LocalMetadataClient.instance()
    end
  end

  def get_url(metadata_name)
    "http://%s/scrapers/metadata_api/%s/%s/" % [ENV["metadata_host"], ENV["SCRAPER_GUID"], CGI.escape(metadata_name)]
  end

  def get_metadata(metadata_name)
    uri = URI.parse(get_url(metadata_name))
    res = Net::HTTP.get_response(uri)
    case res
    when Net::HTTPSuccess
      return JSON.parse(res.body)
    else
      return nil
    end
  end
  
  def post_or_put(method, metadata_name, value, run_id)
    uri = URI.parse(get_url(metadata_name))
    req = (method == :post ? Net::HTTP::Post.new(uri.path) : Net::HTTP::Put.new(uri.path))
    req.set_form_data('run_id'=> run_id,'value'=>JSON.dump([value]))
    res = Net::HTTP.new(uri.host, uri.port).start {|http| http.request(req) }
    case res
    when Net::HTTPSuccess, Net::HTTPRedirection
      return nil
    else
      raise "ERROR SAVING METADATA #{res}"
    end
  end
  
  def get(metadata_name, default = nil)
    metadata = get_metadata(metadata_name)
    metadata ? JSON.parse(metadata['value'])[0] : default
  end
  
  def save(metadata_name, value)
    run_id = ENV['RUNID']
    method = get(metadata_name) ? :put : :post
    post_or_put(method, metadata_name, value,run_id)
  end
end
